src/config.ts

```diff
@@ -1,361 +1,153 @@
-            else if (vimiumgi === false || vimiumgi === "false")
-                set("gimode", "firefox")
-            unset("vimium-gi")
-            set("configversion", "1.1")
-        }
+export async function migrate(cfgObj: object): Promise<object> {
+    let ver = getDeepProperty(cfgObj, ['configversion'])
+    if (!ver) ver = '0.0'
+
+    // Use fall-through to migrate to latest version
+    switch (ver) {
+        case '0.0':
+            await migrate_00_to_10(cfgObj)
+        case '1.0':
+            migrate_10_to_11(cfgObj)
     }
-    if (!get("configversion"))
-        set("configversion", "0.0")
-    while (updaters[get("configversion")] instanceof Function) {
-        await updaters[get("configversion")]()
-    }
-}
 
-/** Read all user configuration from storage API then notify any waiting asynchronous calls
+    setDeepProperty(cfgObj, '1.1', ['configversion'])
+    return cfgObj
+
+    async function migrate_00_to_10(cfgObj: object): Promise<object> {
+        try {
+            // Before we had a config system, we had nmaps, and we put them in the
+            // root namespace because we were young and bold.
+            let legacy_nmaps = await browser.storage.sync.get('nmaps')
+            if (legacy_nmaps['nmaps'])
+                cfgObj['nmaps'] = Object.assign(
+                    legacy_nmaps['nmaps'],
+                    cfgObj['nmaps']
+                )
+        } catch (e) {
+            console.error('Error while migration from 0.0 -> 1.0', e)
+        }
+        return cfgObj
+    }
 
-    asynchronous calls generated by getAsync.
-*/
-async function init() {
-    let syncConfig = await browser.storage.sync.get(CONFIGNAME)
-    schlepp(syncConfig[CONFIGNAME])
-    // Local storage overrides sync
-    let localConfig = await browser.storage.local.get(CONFIGNAME)
-    schlepp(localConfig[CONFIGNAME])
-    await update()
-    INITIALISED = true
-    for (let waiter of WAITERS) {
-        waiter()
+    function migrate_10_to_11(cfgObj: object): object {
+        // Rename vimium-gi => gimode
+        // true => 'nextinput'; false => 'firefox'
+
+        let vimiumgi = getDeepProperty(cfgObj, ['vimium-gi'])
+        if (vimiumgi === true || vimiumgi === 'true')
+            setDeepProperty(cfgObj, 'nextinput', ['gimode'])
+        else if (vimiumgi === false || vimiumgi === 'false')
+            setDeepProperty(cfgObj, 'firefox', ['gimode'])
+        delete cfgObj['vimium-gi']
+        return cfgObj
     }
 }
 
-// Listen for changes to the storage and update the USERCONFIG if appropriate.
-// TODO: BUG! Sync and local storage are merged at startup, but not by this thing.
-browser.storage.onChanged.addListener(
-    (changes, areaname) => {
-        if (CONFIGNAME in changes) {
-            USERCONFIG = changes[CONFIGNAME].newValue
+async function init() {
+    try {
+        const storage = await Utils.getStorage(CONFIG_NAME)
+        const cfg = storage ? storage : DEFAULTS
```

Comment:
Does this deal with https://github.com/cmcaine/tridactyl/pull/93#issuecomment-345971554 at all?

As far as I can tell, it stores all of the defaults in with the user configuration.